lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -1834,6 +1834,147 @@ class ConvertAtenViewAsRealOp : public OpConversionPattern<AtenViewAsRealOp> {
+                  ConversionPatternRewriter &rewriter) const override {
+
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+
+    int64_t offset;
+    if (!matchPattern(op.getOffset(), m_TorchConstantInt(&offset)))
+      return rewriter.notifyMatchFailure(op, "offset must be constant");
+    int64_t dim1;
+    if (!matchPattern(op.getDim1(), m_TorchConstantInt(&dim1)))
+      return rewriter.notifyMatchFailure(op, "dim1 must be constant");
+    int64_t dim2;
+    if (!matchPattern(op.getDim2(), m_TorchConstantInt(&dim2)))
+      return rewriter.notifyMatchFailure(op, "dim2 must be constant");
+
+    Value inputMatrix = adaptor.getSelf();
+    RankedTensorType inType = inputMatrix.getType().cast<RankedTensorType>();
+    int64_t inputRank = inType.getRank();
+    RankedTensorType outType = getTypeConverter()
+                       ->convertType(op->getResult(0).getType())
+                       .cast<RankedTensorType>();
+    Type elementType = inType.getElementType();
+    Location loc = op.getLoc();
+
+    if (inputRank < 2) 
+      return rewriter.notifyMatchFailure(op, "input must have at least two dimensions");
+
+    dim1 = toPositiveDim(dim1, inputRank);
+    if (!isValidDim(dim1, inputRank))
+      return rewriter.notifyMatchFailure(op, "dim1 out of range");
+    dim2 = toPositiveDim(dim2, inputRank);
+    if (!isValidDim(dim2, inputRank))
+      return rewriter.notifyMatchFailure(op, "dim2 out of range");
+    if (dim1 == dim2) 
+      return rewriter.notifyMatchFailure(op, "diagonal dimensions cannot be identical");
+
+    bool offsetNegative = offset < 0;
+    Value dim1Size, dim2Size;
+    dim1Size = getDimOp(rewriter, loc, adaptor.getSelf(), dim1);
+    dim2Size = getDimOp(rewriter, loc, adaptor.getSelf(), dim2);
+    // The pytorch diagonal function yields an empty tensor if the offset is too large or too small.
+    // In these cases diagSize takes on the value 0 below.
+    Value diagSize;
+    if (offsetNegative)
+      diagSize = rewriter.create<arith::MaxSIOp>(
+        loc, 
+        rewriter.create<arith::MinSIOp>(
+          loc, 
+          rewriter.create<arith::AddIOp>(
+            loc, 
+            dim1Size, 
+            rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(offset))
+          ), 
+          dim2Size
+        ), 
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(0))
+      );
+    else
+      diagSize = rewriter.create<arith::MaxSIOp>(
+          loc, 
+          rewriter.create<arith::MinSIOp>(loc, rewriter.create<arith::SubIOp>(
+              loc,
+              dim2Size, 
+              rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(offset))
+          ), 
+          dim1Size), 
+          rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(0))
+      );
+    
+    // TODO: below version also works, need to cast diagSize to Value below
```

Comment:
Remove TODO? Version below seems good to me.

Reply:
In the newer commits, I have deleted the below alternative and just worked with the version, in which computations / case distinctions are made with the rewriter.